### ISSUE TITLE
Protocol extension method added by ExtensionMacro is not visible to compiler

### ISSUE THREAD

**Issue by daniel-hall** [apple/swift#67989]
**Description**

It appears as if extension methods added to protocols using an ExtensionMacro do not become visible to other other code

**Steps to reproduce**

```
// Inside my MacroCompilerPlugin Module

public struct FooExtensionMacro: ExtensionMacro {
    public static func expansion(of node: AttributeSyntax, attachedTo declaration: some DeclGroupSyntax, providingExtensionsOf type: some TypeSyntaxProtocol, conformingTo protocols: [TypeSyntax], in context: some MacroExpansionContext) throws -> [ExtensionDeclSyntax] {
        let decl: DeclSyntax =
        """
        extension Foo {
          var foo: String { "foo" }
          func printFoo() {
            print(foo)
          }
        }
        """
        guard let extensionDecl = decl.as(ExtensionDeclSyntax.self) else {
          return []
        }

        return [extensionDecl]
    }
}

// Inside my MacroLibrary module

@attached(extension, names: named(foo), named(printFoo))
public macro FooExtension() = #externalMacro(module: "MacroCompilerPlugin", type: "FooExtensionMacro")


// Inside the client project / final target

import MacroLibrary

@FooExtension    // <--- apply the macro to generate default property and method
protocol Foo {
  var foo: String { get }
}

/// The above correctly expands to:
/// 
/// protocol Foo {
///   var foo: String { get }
/// }
///
/// extension Foo {
///   var foo: String { "foo" }
///   func printFoo() {
///     print(foo)
///   }
/// }

extension String: Foo { }   // <--- This compiles! Hooray, that means the default implementation of the property foo from the extension macro worked!

"Test".printFoo()   // <--- This doesn't compile ("Value of type ''String" has no member 'printFoo'"). This seems to indicate that the compiler can't "see" the printFoo() extension method from the ExtensionMacro, even though it can see the default implementation of the protocol's foo property from the same macro.
```

**Expected behavior**

```
"Test".printFoo()   // <-- Should compile and print "foo" based on the extension method synthesized by the ExtensionMacro 
```

**Environment**
- Swift: swift-driver version: 1.87.1 Apple Swift version 5.9 (swiftlang-5.9.0.128.2 clang-1500.0.40.1)
- Xcode: version 15.0 beta 6 (15A5219j)
- Deployment target: iOS 17.0

**Additional context**

If I select Refactor -> Inline Macro to add the output of the Extension Macro as actual source code, the extension method works, so this seems to be an issue with visibility or sequencing between macro output and the compiler.

**Comment by DougGregor** [comment]
This is fixed on `main`. Test case added via https://github.com/apple/swift/pull/68059

**Comment by tinder-danielhall** [comment]
Thanks for your work on this @ahoppen and @DougGregor! I was trying this out today on the latest nightly and I think there is one extra wrinkle: the visibility is a problem only in other files (its perfectly visible in the file where the Macro is applied).

If I have all this in the same file it works.

```
import MacroLibrary

@FooExtension    // <--- apply the macro to generate default property and method
protocol Foo {
  var foo: String { get }
}

extension String: Foo { } 

"Test".printFoo()   // prints "foo"
```

But if I have the last line in a separate file, the extension method isn't visible even if I generated it specifically with an `internal` modifier

```
// Some other file that is different from the above
"Test".printFoo()  // <---- Error: "Value of type ''String" has no member 'printFoo'"
```

So perhaps this is a matter of the ExtensionMacro added members only being visible at the fileprivate scope, even if I specifically generate the extension members with the `internal` modifier?

It looks like the test case may have trouble catching this since the extension method is called in the same file where the macro is applied (therefore never triggering the visibility issue). I'm attached a minimal example package here to demonstrate. 

Should this be a new issue or just a further clarification of this existing one? Or is it already fixed but just not in the nightly 5.9 toolchain yet?

[MyMacro.zip](https://github.com/apple/swift/files/12434527/MyMacro.zip)

**Comment by daniel-hall** [comment]
Thanks for your work on this @ahoppen and @DougGregor! I was trying this out today on the latest nightly and I think there is one extra wrinkle: the visibility is a problem only in other files (its perfectly visible in the file where the Macro is applied).

If I have all this in the same file it works.

```
import MacroLibrary

@FooExtension    // <--- apply the macro to generate default property and method
protocol Foo {
  var foo: String { get }
}

extension String: Foo { } 

"Test".printFoo()   // prints "foo"
```

But if I have the last line in a separate file, the extension method isn't visible even if I generated it specifically with an `internal` modifier

```
// Some other file that is different from the above
"Test".printFoo()  // <---- Error: "Value of type ''String" has no member 'printFoo'"
```

So perhaps this is a matter of the ExtensionMacro added members only being visible at the fileprivate scope, even if I specifically generate the extension members with the `internal` modifier?

It looks like the test case may have trouble catching this since the extension method is called in the same file where the macro is applied (therefore never triggering the visibility issue). I'm attached a minimal example package here to demonstrate.

Should this be a new issue or just a further clarification of this existing one? Or is it already fixed but just not in the nightly 5.9 toolchain yet?

[MyMacro.zip](https://github.com/apple/swift/files/12434606/MyMacro.zip)
